### PR TITLE
ReserveNow and CancelReservation Feature

### PR DESCRIPTION
**Pull request by JulianHBuecher**
Hi Jerome,
thank you very much for your immediate response and your guidance regarding this newer version. After polishing and reimplementing my latest features, my pull request for the new simulator version is ready. 
Hopefully I could fulfill your expectations. As guidance for the ReserveNow and CancelReservation feature I have used the OCPP 1.6 Reference Guide from the OCA. If you have further questions, please reach out to me.
Best regards,
Julian

### CODE REVIEW

File: src/assets/station-templates/evlink.station-template.json [hunk 1]
```diff
@@ -40,12 +40,17 @@
       {
         "key": "AuthorizeRemoteTxRequests",
         "readonly": false,
-        "value": "false"
+        "value": "true"
```
Comment (jerome-benoit) [inline src/assets/station-templates/evlink.station-template.json]:
```suggestion
        "value": "false"
```

File: src/charging-station/ChargingStation.ts [hunk 2]
```diff
@@ -132,6 +140,8 @@ export class ChargingStation {
   private readonly sharedLRUCache: SharedLRUCache;
   private webSocketPingSetInterval!: NodeJS.Timeout;
   private readonly chargingStationWorkerBroadcastChannel: ChargingStationWorkerBroadcastChannel;
+  private reservations?: Reservation[];
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]: Reservations are linked to a connector, right? 
So they should be part of the `ConnectorStatus` data structure.

File: src/charging-station/ChargingStation.ts [hunk 3]
```diff
@@ -951,6 +1175,7 @@ export class ChargingStation {
 
   private getStationInfoFromTemplate(): ChargingStationInfo {
     const stationTemplate: ChargingStationTemplate | undefined = this.getTemplateFromFile();
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]:
```suggestion
```

File: src/charging-station/ChargingStation.ts [hunk 4]
```diff
@@ -1066,6 +1291,7 @@ export class ChargingStation {
 
   private initialize(): void {
     const stationTemplate = this.getTemplateFromFile();
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]:
```suggestion
```

File: src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts [hunk 5]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable max-len */
```
Comment (jerome-benoit) [inline src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts]:
```suggestion
```

File: src/utils/Configuration.ts [hunk 6]
```diff
@@ -55,7 +55,9 @@ export class Configuration {
   public static getUIServer(): UIServerConfiguration {
     if (Utils.hasOwnProp(Configuration.getConfig(), 'uiWebSocketServer')) {
       console.error(
-        chalk`{green ${Configuration.logPrefix()}} {red Deprecated configuration section 'uiWebSocketServer' usage. Use 'uiServer' instead}`
+        `${chalk.green(Configuration.logPrefix())} ${chalk.red(
+          "Deprecated configuration section 'uiWebSocketServer' usage. Use 'uiServer' instead"
+        )}`
```
Comment (jerome-benoit) [inline src/utils/Configuration.ts]: ?

File: src/utils/Configuration.ts [hunk 7]
```diff
@@ -137,9 +139,9 @@ export class Configuration {
       (stationTemplateUrl: StationTemplateUrl) => {
         if (!Utils.isUndefined(stationTemplateUrl['numberOfStation'])) {
           console.error(
-            chalk`{green ${Configuration.logPrefix()}} {red Deprecated configuration key 'numberOfStation' usage for template file '${
-              stationTemplateUrl.file
-            }' in 'stationTemplateUrls'. Use 'numberOfStations' instead}`
+            `${chalk.green(Configuration.logPrefix())} ${chalk.red(
+              `Deprecated configuration key 'numberOfStation' usage for template file '${stationTemplateUrl.file}' in 'stationTemplateUrls'. Use 'numberOfStations' instead`
+            )}`
```
Comment (jerome-benoit) [inline src/utils/Configuration.ts]: ?

File: src/utils/Configuration.ts [hunk 8]
```diff
@@ -337,15 +339,19 @@ export class Configuration {
...
     ) {
       console.error(
-        chalk`{green ${Configuration.logPrefix()}} {red Deprecated configuration key '${key}' usage in section '${sectionName}'${
-          logMsgToAppend.trim().length > 0 ? `. ${logMsgToAppend}` : ''
-        }}`
+        `${chalk.green(Configuration.logPrefix())} ${chalk.red(
+          `Deprecated configuration key '${key}' usage in section '${sectionName}'${
+            logMsgToAppend.trim().length > 0 ? `. ${logMsgToAppend}` : ''
+          }`
+        )}`
```
Comment (jerome-benoit) [inline src/utils/Configuration.ts]: ?

File: src/utils/Configuration.ts [hunk 9]
```diff
@@ -337,15 +339,19 @@ export class Configuration {
...
     } else if (!Utils.isUndefined(Configuration.getConfig()[key])) {
       console.error(
-        chalk`{green ${Configuration.logPrefix()}} {red Deprecated configuration key '${key}' usage${
-          logMsgToAppend.trim().length > 0 ? `. ${logMsgToAppend}` : ''
-        }}`
+        `${chalk.green(Configuration.logPrefix())} ${chalk.red(
+          `Deprecated configuration key '${key}' usage${
+            logMsgToAppend.trim().length > 0 ? `. ${logMsgToAppend}` : ''
+          }`
+        )}`
```
Comment (jerome-benoit) [inline src/utils/Configuration.ts]: ?

File: src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts [hunk 10]
```diff
@@ -84,7 +84,17 @@ import {
...
 } from '../../../types';
+import type {
+  OCPP16CancelReservationRequest,
+  OCPP16ReserveNowRequest,
+} from '../../../types/ocpp/1.6/Requests';
+import { ReservationTerminationReason } from '../../../types/ocpp/1.6/Reservation';
+import type {
+  OCPP16CancelReservationResponse,
+  OCPP16ReserveNowResponse,
+} from '../../../types/ocpp/1.6/Responses';
```
Comment (jerome-benoit) [inline src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts]: All types must be exported via `types/index.ts`.

File: src/assets/station-templates/schneider-imredd.station-template.json [hunk 11]
```diff
@@ -45,6 +45,11 @@
         "key": "WebSocketPingInterval",
         "readonly": false,
         "value": "60"
+      },
+      {
+        "key": "ReserveConnectorZeroSupported",
+        "readonly": false,
+        "value": "true"
```
Comment (jerome-benoit) [inline src/assets/station-templates/schneider-imredd.station-template.json]: true or false? Please set consistent  default values for all templates.

File: src/charging-station/ChargingStation.ts [hunk 12]
```diff
@@ -888,6 +901,211 @@ export class ChargingStation {
...
+          RequestCommand.STATUS_NOTIFICATION,
+          OCPPServiceUtils.buildStatusNotificationRequest(
+            this,
+            reservation.connectorId,
+            ConnectorStatusEnum.Available
+          )
+        );
+        break;
+    }
+  }
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts:994]: It looks like reservation lifecycle can be represented via a state machine. Why haven't implemented it as such? 
+ state transitions validation 
+ proper abstraction to trigger action at state change

File: src/charging-station/ChargingStation.ts [hunk 13]
```diff
@@ -888,6 +901,211 @@ export class ChargingStation {
...
+    const reservation = this.getReservationByConnectorId(connectorId);
+    return Utils.isUndefined(reservation) || reservation.idTag !== idTag;
+  }
+
+  public isConnectorReservable(
+    reservationId: number,
+    connectorId?: number,
+    idTag?: string
+  ): boolean {
+    const [alreadyExists, _] = this.doesReservationExists({ id: reservationId });
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]:
```suggestion
    const [alreadyExists] = this.doesReservationExists({ id: reservationId });
```

File: src/charging-station/ChargingStation.ts [hunk 14]
```diff
@@ -888,6 +901,211 @@ export class ChargingStation {
...
+    }
+    const userReservedAlready = Utils.isUndefined(this.getReservationByIdTag(idTag)) ? false : true;
+    const notConnectorZero = Utils.isUndefined(connectorId) ? true : connectorId > 0;
+    const freeConnectorsAvailable = this.getNumberOfReservableConnectors() > 0;
+    return !alreadyExists && !userReservedAlready && notConnectorZero && freeConnectorsAvailable;
+  }
+
+  private getNumberOfReservableConnectors(): number {
+    let reservableConnectors = 0;
+    this.connectors.forEach((connector, id) => {
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]: Please use for ... of loop, they are interruptible.

File: src/types/ocpp/1.6/Responses.ts [hunk 15]
```diff
@@ -109,3 +109,24 @@ export interface OCPP16DataTransferResponse extends JsonObject {
   status: OCPP16DataTransferStatus;
   data?: string;
 }
+
+export enum OCPP16CancelReservationStatus {
```
Comment (jerome-benoit) [inline src/types/ocpp/1.6/Responses.ts]: Please use a type alias to the `GenericStatus` enum.

File: src/types/ocpp/Responses.ts [hunk 16]
```diff
@@ -103,3 +105,15 @@ export const DataTransferStatus = {
   ...OCPP16DataTransferStatus,
 } as const;
 export type DataTransferStatus = OCPP16DataTransferStatus;
+
+export type ReservationStatus = OCPP16ReservationStatus;
+
```
Comment (jerome-benoit) [inline src/types/ocpp/Responses.ts]:
```suggestion
```

File: src/types/ocpp/Responses.ts [hunk 17]
```diff
@@ -103,3 +105,15 @@ export const DataTransferStatus = {
...
+export const ReservationStatus = {
+  ...OCPP16ReservationStatus,
+};
+
+export type CancelReservationStatus = OCPP16CancelReservationStatus;
+
```
Comment (jerome-benoit) [inline src/types/ocpp/Responses.ts]:
```suggestion
```

File: src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts [hunk 18]
```diff
@@ -867,15 +884,24 @@ export class OCPP16IncomingRequestService extends OCPPIncomingRequestService {
           ) === true
         ) {
           connectorStatus.transactionRemoteStarted = true;
+          const startTransactionData: JsonType = {
```
Comment (jerome-benoit) [inline src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts]:
```suggestion
          const startTransactionPayload: JsonType = {
```

And you can strong type to the exact payload type for the OCPP request.

File: src/charging-station/ChargingStation.ts (hunk 2 above)
Comment (JulianHBuecher) [reply src/charging-station/ChargingStation.ts]: Good point, I'll change that

File: src/utils/Configuration.ts (hunk 6 above)
Comment (JulianHBuecher) [reply src/utils/Configuration.ts]: Using the chalk`{...}` syntax, output colouring does not work properly. After changing the syntax to my version, its working fine again. Alternatively I've used ansiColor-package for console colors in the stale version of the simulator.

File: src/utils/Configuration.ts (hunk 7 above)
Comment (JulianHBuecher) [reply src/utils/Configuration.ts]: s.o.

File: src/utils/Configuration.ts (hunk 8 above)
Comment (JulianHBuecher) [reply src/utils/Configuration.ts]: s.o.

File: src/utils/Configuration.ts (hunk 9 above)
Comment (JulianHBuecher) [reply src/utils/Configuration.ts]: s.o.

File: src/assets/station-templates/schneider-imredd.station-template.json (hunk 11 above)
Comment (JulianHBuecher) [reply src/assets/station-templates/schneider-imredd.station-template.json]: I've used different values for simulating stations, which are capable of reserving connector 0 and other ones, which are not capable. What would you prefer? Then I'll change it consistently.

File: src/charging-station/ChargingStation.ts (hunk 2 above)
Comment (jerome-benoit) [reply src/charging-station/ChargingStation.ts]: How OCPP 2.0 handle reservation? If it's at the EVSE level, then it should also be added to `EvseStatus` and helpers must behave accordingly (connectors or evses set in CS template) .

File: src/charging-station/ChargingStation.ts [hunk 19]
```diff
@@ -888,6 +901,211 @@ export class ChargingStation {
...
+    return this.reservations?.find((reservation) => reservation.connectorId === id);
+  }
+
+  public doesReservationExists(reservation: Partial<Reservation>): [boolean, Reservation] {
+    const sameReservation = (r: Reservation) => r.id === reservation.id;
+    const foundReservation = this.reservations?.find(sameReservation);
+    return Utils.isUndefined(foundReservation) ? [false, null] : [true, foundReservation];
+  }
+
+  public async isAuthorized(
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]: I think it should be moved to OCPP16ServiceUtils if possible and renamed to something like `isIdTagAuthorized(): boolean`.
The code seems to rely on local and remote authorization, maybe you can also split the authorization private helpers to `isIdTagLocalAuthorized()`, `isIdTagRemoteAuthorized()` called by the public one. 
And given the boolean output, call site properly build payload, etc.

File: src/types/ChargingStationTemplate.ts [hunk 20]
```diff
@@ -14,6 +14,7 @@ import type {
   MessageTrigger,
   RequestCommand,
 } from './ocpp/Requests';
+import type { Reservation } from './ocpp/Reservation';
```
Comment (jerome-benoit) [inline src/types/ChargingStationTemplate.ts]: Modules export from ocpp directory must done via `ocpp/index.ts`

File: src/assets/station-templates/schneider-imredd.station-template.json (hunk 11 above)
Comment (jerome-benoit) [reply src/assets/station-templates/schneider-imredd.station-template.json]: I doubt the whole charging station reservation makes a lot of sense. What OCPP specs say about that parameter default value?

File: src/charging-station/ocpp/OCPPConstants.ts [hunk 21]
```diff
@@ -101,6 +102,18 @@ export class OCPPConstants {
...
+  static readonly OCPP_RESERVATION_RESPONSE_FAULTED = Object.freeze({ status: ReservationStatus.FAULTED }); // Reservation has not been made, because of connector in FAULTED state
+  static readonly OCPP_RESERVATION_RESPONSE_OCCUPIED = Object.freeze({ status: ReservationStatus.OCCUPIED }); // Reservation has not been made, because all connectors are OCCUPIED
+  static readonly OCPP_RESERVATION_RESPONSE_REJECTED = Object.freeze({ status: ReservationStatus.REJECTED }); // Reservation has not been made, because CS is not configured to accept reservations
+  static readonly OCPP_RESERVATION_RESPONSE_UNAVAILABLE = Object.freeze({ status: ReservationStatus.UNAVAILABLE }); // Reservation has not been made, because connectors are spec. connector is in UNAVAILABLE state
+
+  static readonly OCPP_CANCEL_RESERVATION_RESPONSE_ACCEPTED = Object.freeze({ status: CancelReservationStatus.ACCEPTED }); // Reservation for id has been cancelled has been made
+  static readonly OCPP_CANCEL_RESERVATION_RESPONSE_REJECTED = Object.freeze({ status: CancelReservationStatus.REJECTED }); // Reservation could not be cancelled, because there is no reservation active for id
```
Comment (jerome-benoit) [inline src/charging-station/ocpp/OCPPConstants.ts]: That ones can be aliased to the generic status responses accepted and rejected.

File: src/charging-station/ChargingStation.ts [hunk 22]
```diff
@@ -888,6 +901,211 @@ export class ChargingStation {
     );
   }
 
+  public supportsReservations(): boolean {
+    logger.info(`${this.logPrefix()} Check for reservation support in charging station`);
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]:
```suggestion
    logger.debug(`${this.logPrefix()} Check for reservation support in charging station`);
```

File: src/charging-station/ChargingStation.ts [hunk 23]
```diff
@@ -888,6 +901,211 @@ export class ChargingStation {
...
+    return ChargingStationConfigurationUtils.getConfigurationKey(
+      this,
+      StandardParametersKey.SupportedFeatureProfiles
+    ).value.includes(OCPP16SupportedFeatureProfiles.Reservation);
+  }
+
+  public supportsReservationsOnConnectorId0(): boolean {
+    logger.info(
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]:
```suggestion
    logger.debug(
```

File: src/charging-station/ChargingStation.ts [hunk 24]
```diff
@@ -888,6 +901,211 @@ export class ChargingStation {
...
+        ConnectorStatusEnum.Reserved
+      )
+    );
+  }
+
+  public async removeReservation(
+    reservation: Reservation,
+    reason?: ReservationTerminationReason
+  ): Promise<void> {
+    const sameReservation = (r: Reservation) => r.id === reservation.id;
```
Comment (jerome-benoit) [inline src/charging-station/ChargingStation.ts]: I fail to understand the purpose of that arrow function ...

File: src/charging-station/ChargingStation.ts (hunk 2 above)
Comment (JulianHBuecher) [reply src/charging-station/ChargingStation.ts]: I'd take a look inside the 2.0 spec and as far as I've understood the difference between them is, that you have the possibility to reserve the whole EVSE (by evseId, which is optional). For my use case I'll stay at 1.6 for the first throw.

File: src/assets/station-templates/schneider-imredd.station-template.json (hunk 11 above)
Comment (JulianHBuecher) [reply src/assets/station-templates/schneider-imredd.station-template.json]: According to the spec, it is an optional configuration key. It has not to set

File: src/charging-station/ChargingStation.ts (hunk 2 above)
Comment (jerome-benoit) [reply src/charging-station/ChargingStation.ts]: The simulator code support evses or(exclusif) connectors configuration in the template, so your reservation implementation must support both configurations, independently of the OCPP stack version.

File: src/assets/station-templates/schneider-imredd.station-template.json (hunk 11 above)
Comment (jerome-benoit) [reply src/assets/station-templates/schneider-imredd.station-template.json]: Default to `false` then seems to be the sensible default.

File: src/types/ocpp/1.6/Reservation.ts [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+  expiryDate: Date;
+  idTag: string;
+  parentIdTag?: string;
+}
+
+export enum ReservationTerminationReason {
+  EXPIRED = 'Expired',
+  TRANSACTION_STARTED = 'TransactionStarted',
+  CONNECTOR_STATE_CHANGED = 'ConnectorStateChanged',
+  CANCELED = 'ReservationCanceled',
```
Comment (jerome-benoit) [inline src/types/ocpp/1.6/Reservation.ts]:
```suggestion
  RESERVATION_CANCELED = 'ReservationCanceled',
```

File: src/assets/station-templates/keba.station-template.json [hunk 26]
```diff
@@ -43,6 +43,11 @@
         "key": "WebSocketPingInterval",
         "readonly": false,
         "value": "60"
+      },
+      {
+        "key": "ReserveConnectorZeroSupported",
```
Comment (jerome-benoit) [inline src/assets/station-templates/keba.station-template.json:48]: Is that key part of the OCPP 1.6 standard keys?
If yes, please add it also to the README.md `OCPP-J standard parameters supported` section.

**Pull request merged**